filament/src/materials/tonemapping.mat

```diff
@@ -0,0 +1,387 @@
+        // Vlachos 2016, "Advanced VR Rendering"
+        highp vec3 noise = vec3(dot(vec2(171.0, 231.0), gl_FragCoord.xy + postProcessUniforms.time));
+        noise = fract(noise / vec3(103.0, 71.0, 97.0));
+        // remap from [0..1[ to [-1..1[
+        noise = (noise * 2.0) - 1.0;
+        return vec4(rgba.rgb + (noise / 255.0), rgba.a);
+    }
+
+    vec4 Dither_TriangleNoise(vec4 rgba) {
+        // Gjøl 2016, "Banding in Games: A Noisy Rant"
+        return rgba + triangleNoise(gl_FragCoord.xy * frameUniforms.resolution.zw) / 255.0;
+    }
+
+    vec4 Dither_TriangleNoiseRGB(vec4 rgba) {
+        // Gjøl 2016, "Banding in Games: A Noisy Rant"
+        vec2 uv = gl_FragCoord.xy * frameUniforms.resolution.zw;
+        vec3 dither = vec3(
+                triangleNoise(uv),
+                triangleNoise(uv + 0.1337),
+                triangleNoise(uv + 0.3141)) / 255.0;
+        return vec4(rgba.rgb + dither, rgba.a + dither.x);
+    }
+
+    //------------------------------------------------------------------------------
+    // Dithering dispatch
+    //------------------------------------------------------------------------------
+
+    /**
+     * Dithers the specified RGB color based on the current time and fragment
+     * coordinates the input must be in the final color space (including OECF).
+     * This dithering function assumes we are dithering to an 8-bit target.
+     * This function dithers the alpha channel assuming premultiplied output
+     */
+    vec4 dither(vec4 rgba) {
+#if DITHERING_OPERATOR == DITHERING_NONE
+        return rgba;
+#elif DITHERING_OPERATOR == DITHERING_INTERLEAVED_NOISE
+        return Dither_InterleavedGradientNoise(rgba);
+#elif DITHERING_OPERATOR == DITHERING_VLACHOS
+        return Dither_Vlachos(rgba);
+#elif DITHERING_OPERATOR  == DITHERING_TRIANGLE_NOISE
+        return Dither_TriangleNoise(rgba);
+#elif DITHERING_OPERATOR  == DITHERING_TRIANGLE_NOISE_RGB
+        return Dither_TriangleNoiseRGB(rgba);
+#endif
+    }
+
+    vec3 resolveFragment(const ivec2 uv) {
+        return texelFetch(materialParams_colorBuffer, uv, 0).rgb;
+    }
+
+    vec4 resolveAlphaFragment(const ivec2 uv) {
+        return texelFetch(materialParams_colorBuffer, uv, 0);
+    }
+
+    vec4 resolve() {
+#if POST_PROCESS_OPAQUE
+        vec4 color = vec4(resolveFragment(ivec2(getUV())), 1.0);
+        color.rgb  = tonemap(color.rgb);
+        color.rgb  = OECF(color.rgb);
+        color.a    = luminance(color.rgb);
+#else
+        vec4 color = resolveAlphaFragment(ivec2(getUV()));
+        color.rgb /= color.a + FLT_EPS;
+        color.rgb  = tonemap(color.rgb);
+        color.rgb  = OECF(color.rgb);
+        color.rgb *= color.a + FLT_EPS;
+#endif
+        return color;
+    }
+
+    void postProcess(inout PostProcessInputs postProcess) {
+        postProcess.color = resolve();
+        if (materialParams.dithering > 0) {
+            postProcess.color = dither(postProcess.color);
+        }
+    }
```

Comment:
It's quite annoying that we have to put all those files together in a single one. Maybe we should add an #include capability to matc/filamat (glslang can handle it) so we could keep all that code in `shaders/`.

Reply:
Yeah I agree. I'll look into using #include.